staging/src/k8s.io/mount-utils/mount_linux.go

```diff
@@ -637,6 +638,46 @@ func SearchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
 	return refs, nil
 }
 
+// IsMountPoint determines if a directory is a mountpoint.
+// It should return ErrNotExist when the directory does not exist.
+// IsMountPoint is more expensive than IsLikelyNotMountPoint.
+// IsMountPoint detects bind mounts in linux.
+// IsMountPoint enumerates all the mountpoints using List() and
+// the list of mountpoints may be large, then it uses
+// isMountPointMatch to evaluate whether the directory is a mountpoint.
+func (mounter *Mounter) IsMountPoint(file string) (bool, error) {
+	isMnt, sure, isMntErr := mountinfo.MountedFast(file)
```

Comment:
before calling MountedFast, should we still call IsLikelyNotMountPoint like before so it can solve certain cases first.

Also if file does not exist, it will return immediately instead of executing the following.

Reply:
MountedFast already uses stat function internally. I would like to keep this as it is backed by actual tests in the moby/sys CI; so bug fixes in the upstream library can be shipped with more confidence.

This change is backward compatible when file does not exist, it returns an error ErrNotExist and the user is supposed to determine what to do with it. This is where: https://github.com/kubernetes/kubernetes/pull/109217/files#diff-5a53703b1baddb2caa650077f26075c59361575c84387e1e9ad5570a738dd473L338-L341, ErrNotExist is returned. This is the comment (https://github.com/kubernetes/kubernetes/pull/109217/files#diff-e0ed3eecf397212be5c090559125652c439c5b93699be999c4bf7e06ae6531a2L224) that mentions ErrNotExist should be handled by the caller.

Reply:
the the first link the correct one you mentioned? Just make sure mountinfo.MountedFast(file) returns correct error if file does not exit. Looks like it is using open sys call.

Reply:
Technically, MountedFast calls stat first, but it works in all cases: https://go.dev/play/p/GyieSQgECO-.

Reply:
Before IsNotMountPoint() function implementation, it calls isLikelyNotMountPoint first. Is there any downside of using/not using this function first?

Reply:
Really good question; mountedFast has the same logic: 
MountedFast first normalizes the path and evaluates all the symlinks first. Lstat is used (instead of stat in case of IsLikelyNotMountPoint). With Lstat, symlinks are not followed; however since the path is already normalized it is not required.

The other part of the logic is the same: https://github.com/moby/sys/blob/523ebb9c6cc361d558aaf903946709d78efa0f6a/mountinfo/mounted_unix.go#L13-L31